kombu/common.py

```diff
@@ -105,6 +105,12 @@ def declaration_cached(entity, channel):
 
 def maybe_declare(entity, channel=None, retry=False, **retry_policy):
     """Declare entity (cached)."""
+    if retry:
+        return _imaybe_declare(entity, channel, **retry_policy)
+    return _maybe_declare(entity, channel)
+
+
+def _maybe_declare(entity, channel):
     is_bound = entity.is_bound
```

Comment:
@thedrow when `maybe_declare` gets `channel=None` (the default), it will take the channel from `entity`, save it to a local var, then pass it to `_maybe_declare` though the `ensure` func.
If the connection is lost, `ensure` opens a new connection, and calls `revive(entity)` which sets `entity.channel` and retries. However because `channel` was passed directly to `_maybe_declare` it won't use the new channel.
This change makes sure that `_maybe_declare` always uses the most up-to-date channel.